mkfs.go

```diff
@@ -0,0 +1,134 @@
+		}
+		log.Println("No /perm mountpoint found.")
+	}
+	return nil
+}
+
+// makeFS runs bcachefs format on the `perm` rootdev partition
+func makeFS() error {
+	dev := rootdev.Partition(rootdev.Perm)
+	log.Printf("Creating file system on %s", dev)
+
+	//make temp dir
+	tmp, err := os.MkdirTemp("", "gokrazy-mkfs.bcachefs-")
+	if err != nil {
+		return err
+	}
+	defer os.RemoveAll(tmp)
+	log.Printf("Writing self-contained bcachefs-tools to %s", tmp)
+
+	// write each file from embedded dir
+	dirents, err := bcachefsEmbedded.ReadDir(bcachefsRoot)
+	for _, d := range dirents {
+		b, err := bcachefsEmbedded.ReadFile(filepath.Join(bcachefsRoot, d.Name()))
+		if err != nil {
+			return err
+		}
+		fn := filepath.Join(tmp, d.Name())
+		if err := os.WriteFile(fn, b, 755); err != nil {
+			return err
+		}
+	}
+
+	// exec bcachefs format on rootdev partition
+	mkfs := exec.Command(filepath.Join(tmp, "ld-linux-aarch64.so.1"), filepath.Join(tmp, "bcachefs"), dev)
+	mkfs.Env = append(os.Environ(), "LD_LIBRARY_PATH="+tmp)
+	mkfs.Stdout = os.Stdout
+	mkfs.Stderr = os.Stderr
+	log.Printf("exec bcachefs with args: %v", mkfs.Args)
+	if err := mkfs.Run(); err != nil {
+		return fmt.Errorf("%v: %v", mkfs.Args, err)
+	}
+	log.Printf("Success formatting rootdev perm partition!")
+	return nil
+}
+
+// reboot triggers a reboot so that /perm will be picked up and mounted on next
+// boot. It is pointless to try and mount the file system here from within this
+// process, as gokrazy services are run in a separate mount namespace.
+func reboot() error {
+	log.Printf("Rebooting system so /perm is mounted on boot.")
+
+	httpPassword, err := readConfigFile("gokr-pw.txt")
+	if err != nil {
+		return fmt.Errorf("could read neither /perm/gokr-pw.txt, nor /etc/gokr-pw.txt, nor /gokr-pw.txt: %v", err)
+	}
+	port, err := ioutil.ReadFile("/etc/http-port.txt")
+	if err != nil {
+		return err
+	}
+	req, err := http.NewRequest("POST", "http://localhost:"+string(port)+"/reboot", nil)
+	if err != nil {
+		return err
+	}
+	req.Header.Set("Authorization", "Basic "+base64.StdEncoding.EncodeToString([]byte("gokrazy:"+httpPassword)))
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return err
+	}
+	if got, want := resp.StatusCode, http.StatusOK; got != want {
+		return fmt.Errorf("rebooting device: unexpected HTTP status code: got %d, want %d", got, want)
+	}
+	return nil
+}
+
+// readConfigFile reads configuration files from /perm /etc or / and returns
+// trimmed content as string.
+//
+// TODO: de-duplicate this with gokrazy.go into a gokrazy/internal package
+func readConfigFile(fileName string) (string, error) {
+	str, err := ioutil.ReadFile("/perm/" + fileName)
+	if err != nil {
+		str, err = ioutil.ReadFile("/etc/" + fileName)
+	}
+	if err != nil && os.IsNotExist(err) {
+		str, err = ioutil.ReadFile("/" + fileName)
+	}
+	return strings.TrimSpace(string(str)), err
+}
+
+func main() {
+	if err := checkExistingFS(); err != nil {
```

Comment:
You have restructured the program such that it also reboots even when not creating a file system. This makes it impossible to include the mkfs.bcachefs package in gokrazy installations. `gok run` still works, but adding the package permanently should also work.

(Also, it tries to create the file system even when one is already mounted.)

Can you revert your changes to mkfs.go as much as possible please? That will also make ongoing maintenance easier.